applications/StructuralMechanicsApplication/custom_utilities/multipoint_constraint_data.hpp

```diff
@@ -0,0 +1,246 @@
+    {
+        return this->slaveDofMasterDataMap[slaveDOF].masterWeights;
+    }
+
+    /**
+		Get the Total number of MasterDOFs in the current Node
+		@return Total number of MasterDOFs in the current Node
+		 */
+    const int GetTotalNumbeOfMasterDOFs()
+    {
+        this->totalNumberOfMasterDOFs = 0;
+        for (const auto it : this->slaveDofMasterDataMap)
+        {
+            this->totalNumberOfMasterDOFs += (it.second).masterDOFs.size();
+        }
+        return this->totalNumberOfMasterDOFs;
+    }
+
+    /**
+		Get the SlaveDOFs vector for this Node
+		@return SlaveDOFs vector for this Node
+		*/
+    const std::vector<DofPointerType> GetSlaveDOFs()
+    {
+        return this->slaveDofVector;
+    }
+
+    ///@
+
+    ///@name Static Operations
+    ///@{
+    /**
+		 * Returns the string containing a detailed description of this object.
+		 * @return the string with informations
+		 */
+    virtual void GetInfo() const
+    {
+        std::cout << std::endl;
+        std::cout << "===============================================================" << std::endl;
+        std::cout << "Total number of slaves :: " << (slaveDofVector).size() << std::endl;
+        for (const auto it : this->slaveDofMasterDataMap)
+        {
+            std::cout << "SlaveDOF :: " << (it.first)->EquationId() << std::endl;
+            for (unsigned int i = 0; i < (it.second).masterDOFs.size(); ++i)
+            {
+                std::cout << "\t MasterDOFs :: " << ((it.second).masterDOFs[i])->EquationId() << std::endl;
+                std::cout << "\t MasterWeights ::" << (it.second).masterWeights[i] << std::endl;
+            }
+        }
+        std::cout << "===============================================================" << std::endl;
+        std::cout << std::endl;
+    }
+
+    ///@}
+    virtual void PrintInfo(std::ostream &rOStream) const
+    {
+        rOStream << " MpcData object " << std::endl;
+    }
+
+  private:
+    ///@name Member Variables
+    ///@{
+    std::vector<DofPointerType> slaveDofVector;
+    unsigned int totalNumberOfMasterDOFs;
+
+    struct masterData
+    {
+        std::vector<DofPointerType> masterDOFs;
+        std::vector<double> masterWeights;
+    };
+
+    std::map<DofPointerType, masterData> slaveDofMasterDataMap;
+    ///@}
+
+    ///@name Serialization
+    ///@{
+    friend class Serializer;
+
+    virtual void save(Serializer &rSerializer) const
+    {
+        KRATOS_SERIALIZE_SAVE_BASE_CLASS(rSerializer, Flags);
+    }
+
+    virtual void load(Serializer &rSerializer)
+    {
+        KRATOS_SERIALIZE_LOAD_BASE_CLASS(rSerializer, Flags);
+    }
+
+    ///@}
+};
+
+///@name Input/Output funcitons
+///@{
+
+inline std::istream &operator>>(std::istream &rIStream, MpcData &rThis);
+
+inline std::ostream &operator<<(std::ostream &rOStream, const MpcData &rThis)
+{
+    return rOStream;
+}
+
+///@}
+
+} // namespace Kratos
+
+#endif // CONSTRAINT_SLAVE_H_INCLUDED

```

Comment:
@RiccardoRossi  Passing geometry is also ok (not element or condition). 

About having one global map ... what happens when running parallel ? 
Also I am confused and cannot understand how should a scheme or builder and solver know how to modify a specific element matrix ? In that case we should also save the slave DOFs on the node ? 

About having multiple stages : 
I think it will not be a problem. This is because, (please correct me if I am wrong)

- DOFs are Unique and the equation IDs that they will have also are unique even for Fractional step solver ?? (If not below points are not valid). 
- Then the procedure of applying MPC will not find this slave equation ID in the provided equationIDs and nothing will happen (this is like implicitly finding active MPC). I would like to really test this for fractional step solver. But I just saw that Fractional step strategy always uses ResidualBasedEliminationBuilderAndSolver for pressure and velocity for which I did not implement the MPC.